In [1]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.svm import SVC
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import numpy as np
import time

### For this benchmarking experiment, we will utilize the Iris Dataset

In [2]:
# Load data

iris = load_iris()
X, y = iris.data, iris.target

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### Model Definition

In [3]:
model = SVC()

### Grid Search Application

In [4]:
# Parameters for Grid Search
param_grid = {'C': [0.1, 1, 10, 100], 'gamma': [1, 0.1, 0.01, 0.001], 'kernel': ['rbf', 'poly', 'sigmoid']}

# Start the timer for Grid Search
start_time = time.time()

# Execute Grid Search
grid_search = GridSearchCV(model, param_grid, refit=True, verbose=0)
grid_search.fit(X_train, y_train)

# Calculate the execution time of Grid Search
grid_search_time = time.time() - start_time

print("Best score of Grid Search: ", grid_search.best_score_)
print("Processing time of Grid Search: {:.2f} seconds".format(grid_search_time))

Best score of Grid Search:  0.9714285714285715
Processing time of Grid Search: 0.96 seconds


### Random Search Application

In [5]:
# Parameters for Random Search
param_distributions = {'C': np.logspace(-2, 2, 100), 'gamma': np.logspace(-3, 1, 100), 'kernel': ['rbf', 'poly', 'sigmoid']}

# Start the timer for Random Search
start_time = time.time()

# Execute Random Search
random_search = RandomizedSearchCV(model, param_distributions, n_iter=50, refit=True, verbose=0, random_state=42)
random_search.fit(X_train, y_train)

# Calculate the execution time of Random Search
random_search_time = time.time() - start_time


print("Best score of Random Search: ", random_search.best_score_)
print("Processing time of Random Search: {:.2f} seconds".format(random_search_time))

Best score of Random Search:  0.9619047619047618
Processing time of Random Search: 0.55 seconds


## Results and Discussion

The benchmarking experiment results demonstrate key differences between the Grid Search and Random Search algorithms in terms of performance and efficiency.

Firstly, the Grid Search algorithm achieved a slightly higher score of 0.9714 compared to the Random Search's 0.96. This difference, in percentage terms, is approximately 1.19%. While this indicates a marginally better performance by the Grid Search in accuracy, the difference is relatively small.

However, when considering processing time, the distinction becomes more significant. The Grid Search took 0.96 seconds, whereas the Random Search completed in just 0.55 seconds. This means that Random Search was approximately 42.7% faster than Grid Search. In scenarios where processing speed is crucial, such as in real-time applications, this difference in time efficiency can be highly impactful.

While both algorithms performed well, Random Search stands out as the more efficient option in this experiment. Its ability to achieve nearly similar accuracy to Grid Search but in significantly less processing time underlines its suitability for situations where quick decision-making is essential. This advantage can be particularly valuable in real-world applications where computational resources and time are critical factors.